cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -564,7 +564,7 @@ struct GroupedReducingAggregator : public GroupedAggregator {
   template <typename T = Type>
   static enable_if_decimal<T, std::shared_ptr<DataType>> GetOutType(
       const std::shared_ptr<DataType>& in_type) {
-    return in_type;
+    return WidenDecimalToMaxPrecision(in_type).ValueOrDie();
```

Comment:
I don't think we should widen the output type for all the subclasses of `GroupedReducingAggregator`. Promoting to max precision is reasonable for `sum` but is not necessarily the case for others.

Reply:
Yeah, different databases do different things. E.g., for a mean, duckdb converts to a double (which is also what the array aggregate does for non-decimal numeric types in arrow). Do you have a preference for what the type for the other aggregates _should_ be?